### ISSUE TITLE
[bitnami/etcd] User not found/There is no user created for TLS client security model

### ISSUE THREAD

**Issue by Venthe** [bitnami/charts#8920]
### Name and Version

bitnami/etcd 6.13.2

### What steps will reproduce the bug?

```bash
helm upgrade \
    --install \
    --version=6.13.2 \
    --create-namespace \
    --values=etcd-values.yml \
    --namespace=external-dns \
    etcd bitnami/etcd
```

With certs generated as such:
```bash
#!/usr/bin/env bash

NAMESPACE=${NAMESPACE:-external-dns}

 function jq() {
     docker run \
         --rm \
         --interactive \
         docker.io/apteno/alpine-jq:2022-02-01 \
         jq "${@}"
 }

function yq() {
  docker run \
    --rm \
    --interactive \
    --volume "${PWD}":/workdir \
    mikefarah/yq:4.18.1 \
    "$@"
}

function _cfssl() {
    local cmd=$1
    shift
    docker run \
        --rm \
        --tty=${TTY:-false} \
        --volume "${PWD}:/workdir" \
        --interactive \
        --entrypoint="${cmd}" \
        docker.io/cfssl/cfssl:1.6.1 \
        "${@}"
}

function cfssl() {
    _cfssl cfssl "${@}"
}

function cfssljson() {
    _cfssl cfssljson "${@}"
}

function read_csr() {
    openssl req -text -noout -verify -in "$1.csr"
}

function read_pem() {
    openssl x509 -in $1.pem -noout -text
}

function generate() {
    cfssl gencert \
        -ca=ca.pem \
        -ca-key=ca-key.pem \
        -config=ca-config.modified.json \
        -profile=$1 \
        -hostname=$2 \
        - \
        | cfssljson -bare $3
}

function validate() {
    read_csr $1
    read_pem $1
}

function pem_to_crt() {
    openssl x509 -outform der -in $1.pem -out $2.crt
}

function PKCS1_to_PKCS8() {
    openssl pkcs8 -topk8 -nocrypt -in ./$1.pem -out $2.key
}

if [[ "$#" -ne 0 ]]; then
  TTY=true _cfssl $@
  exit $?
fi

rm *.csr *.pem *.json *.crt

cfssl print-defaults config | tee ca-config.json
cfssl print-defaults csr | tee ca-csr.json

# 87600h -> 10 years
cat ca-config.json | jq ". \
    | .signing.default.expiry = \"87600h\" \
    | del(.signing.profiles.www) \
    | .signing.profiles.server = {\"expiry\":\"87600h\", \"usages\": [\"signing\",\"key encipherment\",\"server auth\"]} \
    | .signing.profiles.client = {\"expiry\":\"87600h\", \"usages\": [\"signing\",\"key encipherment\",\"client auth\"]} \
    | .signing.profiles.peer = {\"expiry\":\"87600h\", \"usages\": [\"signing\",\"key encipherment\",\"server auth\",\"client auth\"]} \
    " \
    | tee ca-config.modified.json

cat ca-csr.json | jq ". \
    | .CN = \"Local cluster ETCD\" \
    | .key.algo=\"rsa\" | .key.size=2048 \
    | .names[0].\"C\" = \"PL\" \
    | .names[0].\"ST\" = \"Masovia\" \
    | .names[0].\"L\" = \"Warsaw\" \
    | del(.hosts) \
    " \
    | tee ca-csr.modified.json

cfssl gencert -initca ca-csr.modified.json | cfssljson -bare ca -

echo '{"CN":"etcd","hosts":[""],"key":{"algo":"rsa","size":2048}}' \
    | generate peer "etcd,etcd-headless,localhost,*.etcd-headless.external-dns.svc.cluster.local" client

kubectl delete secret/etcd-client-tls --namespace=$NAMESPACE
kubectl create secret generic \
    --from-file=client.pem=./client.pem \
    --from-file=ca.pem=./ca.pem \
    --from-file=client-key.pem=./client-key.pem \
    --namespace=$NAMESPACE \
    etcd-client-tls
```

Then, while logged in to etcd:
```bash
I have no name!@etcd-0:/opt/bitnami/etcd$ etcdctl user list --cert $ETCD_CERT_FILE --key $ETCD_KEY_FILE --cacert $ETCD_TRUSTED_CA_FILE --endpoints=https://etcd:2379 --user root --password BDvQMmluQy
root
I have no name!@etcd-0:/opt/bitnami/etcd$ etcdctl user list --cert $ETCD_CERT_FILE --key $ETCD_KEY_FILE --cacert $ETCD_TRUSTED_CA_FILE --endpoints=https://etcd:2379
{"level":"warn","ts":"2022-02-07T02:17:13.565Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003d2c40/etcd:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: user name not found"}
Error: etcdserver: user name not found
```

### Are you using any custom parameters or values?

```
auth:
  rbac:
    enabled: true
   rootPassword: BDvQMmluQy
  client:
    enableAuthentication: true
    secureTransport: true
    certFilename: client.pem
    certKeyFilename: client-key.pem
    caFilename: ca.pem
    existingSecret: etcd-client-tls
# image.debug: true
extraEnvVars:
  - name: ETCD_LOG_LEVEL
    value: debug
```

### What is the expected behavior?

I am able to use ETCD (in my case - external dns & coredns with TLS certs), or `etcdctl`

### What do you see instead?

No user

### Additional information

```bash
# MANUALLY CREATE USER. Forgot to copy first line
I have no name!@etcd-0:/opt/bitnami/etcd$ etcdctl --cert $ETCD_CERT_FILE --key $ETCD_KEY_FILE --cacert $ETCD_TRUSTED_CA_FILE --endpoints=https://etcd:2379 --user root --password wn8j4MVkcM user get root
User: root
Roles: root
I have no name!@etcd-0:/opt/bitnami/etcd$ etcdctl --cert $ETCD_CERT_FILE --key $ETCD_KEY_FILE --cacert $ETCD_TRUSTED_CA_FILE --endpoints=https://etcd:2379 --user root --password wn8j4MVkcM user grant-role etcd root
Role root is granted to user etcd
I have no name!@etcd-0:/opt/bitnami/etcd$ etcdctl user list --cert $ETCD_CERT_FILE --key $ETCD_KEY_FILE --cacert $ETCD_TRUSTED_CA_FILE --endpoints=https://etcd:2379
etcd
root
```

**Comment by Venthe** [comment]
Excuse me being laconic; I'm at a second day fighting with this...
If anything is unclear, please ping me

**Comment by migruiz4** [comment]
Hi @Venthe,

I haven't been able to reproduce your issue.

Generated the certificates and the secret using your script, then installed the Helm chart using:
```console
helm install --version 6.13.2 -f ./values.yml --namespace external-dns etcd bitnami/etcd
```

With the following `values.yaml`:
```yaml
auth:
  rbac:
    enabled: true
    rootPassword: "BDvQMmluQy"
  client:
    enableAuthentication: true
    secureTransport: true
    certFilename: "client.pem"
    certKeyFilename: "client-key.pem"
    caFilename: "ca.pem"
    existingSecret: "etcd-client-tls"
image.debug: true
extraEnvVars:
  - name: "ETCD_LOG_LEVEL"
    value: "debug"
```

Without any additional action got the following output:
```console
I have no name!@etcd-0:/opt/bitnami/etcd$ . /opt/bitnami/scripts/etcd-env.sh
I have no name!@etcd-0:/opt/bitnami/etcd$ etcdctl user list --user root:BDvQMmluQy --cert="$ETCD_CERT_FILE" --key="$ETCD_KEY_FILE" --cacert="$ETCD_TRUSTED_CA_FILE" --endpoints=https://etcd:2379
root
I have no name!@etcd-0:/opt/bitnami/etcd$ etcdctl user get root --user root:BDvQMmluQy --cert="$ETCD_CERT_FILE" --key="$ETCD_KEY_FILE" --cacert="$ETCD_TRUSTED_CA_FILE" --endpoints=https://etcd:2379
User: root
Roles: root
```
 
I think this issue may be caused by one of the following reasons:
- The `values.yaml` is not properly formatted. Probably not the case, but could be since I had to add missing quotes and fix indentations in the example provided.
- Helm is being deployed using the PVC from a previous installation. Although you remove the deployment using Helm, PVC will remain in the cluster and, if the same release name is used, the same PVC will be attached the next time the chart is installed, causing Etcd to use the previously persisted data.

  ```
  $ helm uninstall etcd -n external-dns
  release "etcd" uninstalled
  $ kubectl get pvc -n external-dns
  NAME          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
  data-etcd-0   Bound    pvc-f1c5b9ec-7ab1-4937-8a06-76cb24c13b8d   8Gi        RWO            hostpath       16m
  $ kubectl delete pvc -n external-dns data-etcd-0 
  persistentvolumeclaim "data-etcd-0" deleted
  ```

In addition, it would be helpful if you could share the logs of your etcd pod: `kubectl logs -n external-dns etcd-0`. That would help us determine why the `root` user is not being created.

As reference, this is the expected output:
```
$ kubectl logs -n external-dns etcd-0
etcd 13:25:05.22 
etcd 13:25:05.22 Welcome to the Bitnami etcd container
etcd 13:25:05.23 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-etcd
etcd 13:25:05.23 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-etcd/issues
etcd 13:25:05.23 
etcd 13:25:05.23 INFO  ==> ** Starting etcd setup **
etcd 13:25:05.25 INFO  ==> Validating settings in ETCD_* env vars..
etcd 13:25:05.25 INFO  ==> Initializing etcd
etcd 13:25:05.25 INFO  ==> Generating etcd config file using env variables
etcd 13:25:05.26 INFO  ==> There is no data from previous deployments
etcd 13:25:05.26 INFO  ==> Enabling etcd authentication
etcd 13:25:05.27 INFO  ==> Starting etcd in background
etcd 13:25:08.53 INFO  ==> Stopping etcd
Terminated

etcd 13:25:09.54 INFO  ==> ** etcd setup finished! **
etcd 13:25:09.59 INFO  ==> ** Starting etcd **
```

The root user is generated at `Enabling etcd authentication` : https://github.com/bitnami/bitnami-docker-etcd/blob/8cc16141f4d7908839cafdc4688cd98fa56f77e1/3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh#L333-L346

**Comment by Venthe** [comment]
I think I might not have started the problem clearly. As far as I
understand how TLS authentication should work in etcd, the user should be
able to authenticate via certificate alone - username is provided with the
cert as CN if I remember correctly.

As such, no client can actually connect via certs alone, because cert is
not signed as root. I do believe though that this worked some time ago,
before I've pinned the version.

As for pvc, I've nuked CSI a few times, testing with two and even nuking
the cluster. I'll verify with the linted yaml, though they work imo,
because you can see both the certs and env values.

To reiterate : you cannot do any action with certs alone because user with
the name specified in cert had not been created

**Comment by migruiz4** [comment]
Authentication using only certs also works, but it is necessary to create an additional key pair for the `root` user and sign it with the same CA. According to ETCD documentation, the cert needs the CN to match the username. From https://etcd.io/docs/v3.5/op-guide/security/ : 

--client-cert-auth: When this is set etcd will check all incoming HTTPS requests for a client certificate signed by the trusted CA, requests that don’t supply a valid client certificate will fail. If [authentication](https://etcd.io/docs/v3.5/op-guide/authentication/) is enabled, the certificate provides credentials for the user name given by the Common Name field.

This variable is set when using `auth.client.enableAuthentication=true`

I did the following:
```console
#Create root-client key
$ openssl genrsa -out client-root.key 2048
#Create root-client csr with CN=root
$ openssl req -new -sha256 -key client-root.key -subj "/CN=root/" -out client-root.csr
#Sign it with the same CA used by ETCD
$ openssl x509 -req -in client-root.csr -CA ca.pem -CAkey ca-key.pem -CAcreateserial -out client-root.crt -days 500 -sha256
```

Then, create a new secret that will be used by the etcd client:
```
kubectl create secret generic \
    --from-file=client.pem=./client-root.crt \
    --from-file=ca.pem=./ca.pem \
    --from-file=client-key.pem=./client-root.key \
    --namespace=external-dns \
    etcd-client-root
```

And create a etcd-client pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    run: etcd-client
  name: etcd-client
  namespace: external-dns
spec:
  securityContext:
    fsGroup: 1001
  containers:
    - command:
        - sleep
        - infinity
      image: docker.io/bitnami/etcd:3.5.1-debian-10-r88
      name: etcd-client
      resources: {}
      volumeMounts:
        - name: certs
          mountPath: /certs
  volumes:
  - name: certs
    secret:
      secretName: etcd-client-root
      defaultMode: 256
  dnsPolicy: ClusterFirst
  restartPolicy: Never
status: {}
```

Log into the pod, and check you can authenticate using only certificates:
```console
$ k exec -n external-dns -it etcd-client -- bash
I have no name!@etcd-client:/opt/bitnami/etcd$  etcdctl user list --cacert="/certs/ca.pem" --cert="/certs/client.pem" --key="/certs/client-key.pem" --endpoints=https://etcd:2379
root
```

To clarify, the certificates you generated with the script in the description are used by the ETCD server to configure the client-to-server communication.
But for client-cert-auth to work, additional certificates need to be created for each client with Common Name (CN) matching its username.

**Comment by Venthe** [comment]
Please correct me if I'm wrong - did at any point etcd create users for client cert? I am ALMOST sure that it did, but if not, we can close this bug report

**Comment by migruiz4** [comment]
Hi @Venthe, 

I would like to clarify some concepts around ETCD TLS settings.

ETCD has two communication channels you can secure with TLS in an ETCD cluster: 
- Server-to-server (aka peer-to-peer).
- Client-to-server.

Each channel has its own certificates, so they don't need to use the same certification authority (CA).

When we refer to `client certs`, it could be confusing because we do not refer to the actual client certificates, but the server-side certificates for the client-to-server communication.

In a client-to-server communication, the client won't need its own certificate but just the CA public cert to validate the server CN or AltNames, and encryption to establish an HTTPS channel.

But the setting `client-cert-auth`, value `auth.client.enableAuthentication` of the chart, allows users to authenticate using a key-pair, signed by the same CA used to generate the server-side certificates, but with its username as Common Name (CN).

The chart has never created client-side client-to-server certificates, but server-side certs can be autogenerated using the value `auth.client.useAutoTLS`, which enables the etcd `--auto-tls` setting (more info in the [docs](https://etcd.io/docs/v3.5/op-guide/security/#example-4-automatic-self-signed-transport-security)). 

So, to sum things up, if you'd like to secure your client-to-server communications with `client-cert-auth`, you will need to create the following certificates:
```
# CA certs
ca.key <--- This one has to be kept safe
ca.crt <--- This one is needed by every client

# Server-side certs
etcd-server.crt <--|
etcd-server.key <---- These two + ca.crt go into the client-certs secret.

# Certs for your users to log in, needed by your applications, not the etcd server
user1.crt 
user1.key
---
user2crt
user2.key
---
...
```

I hope this resolve this question and helps other users running into the same issue.

**Comment by Venthe** [comment]
> The chart has never created client-side client-to-server certificates

Thank you for clarification. I'm closing the issue then

Sorry for taking you up for a goose-chase, and taking up your time. Considering what you wrote, I believe that I made it work originally by first starting with autogenerated TLS, then during test I have to had switched certs to mine, which would allow me to connect my clients to it.

In my case, I'm actually fully content with certs being root capable, as my etcd is used in a toy cluster for ephemeral DNS resolution.

_Completely OT:_ I am actually moving off from original, operator based chart - my whole action here was to provide certs for two clients which used to be done by an operator. Do you think that this would fit your chart? To be more precise: **can you store autogenerated client keys in a secret, provisioning an user for them as well?** if so, I'd create a feature request.

Either way, thank you again; you were a real help!

**Comment by migruiz4** [comment]
> Completely OT: I am actually moving off from original, operator based chart - my whole action here was to provide certs for two clients which used to be done by an operator. Do you think that this would fit your chart? To be more precise: can you store autogenerated client keys in a secret, provisioning an user for them as well? if so, I'd create a feature request.

I'm sorry but I don't think it would be possible to add that feature nor do I think it would fit in the chart. 
Creating users and provisioning would be part of the day-to-day Etcd operations, while Helm charts target is to deploy and provision applications. 

So features that are not day-zero operations would be out of the scope of the helm chart and require other tools.
